src/HooglePlus/FilterTest.hs

```diff
@@ -113,80 +110,81 @@ mapRandomMultiple env count = do
         replace _ x = x
 
 eval_ :: [String] -> String -> Int -> IO (Maybe (Either InterpreterError String))
-eval_ modules line time = timeout time $ runInterpreter $ do {
-  setImports modules;
+eval_ modules line time = timeout time $ runInterpreter $ do
+  setImports modules
   eval line
-}
 
 runChecks :: Environment -> RType -> UProgram -> IO Bool
 runChecks env goalType prog = do
   list <- replicateM 10 (runChecks' modules funcSig body)
   return $ or list
   where
-    args = _arguments env
-    modules = "Prelude" : Set.toList (_included_modules env)
-    argList = Map.toList args
-    argNames = map fst argList
-    argTypes = map snd argList
-    monoGoals = map toMonotype argTypes
-    funcSig = mkFunctionSigStr (monoGoals ++ [goalType])
-    body = mkLambdaStr argNames prog
+    (modules, funcSig, body, _) = extractSolution env goalType prog
 
 runChecks' :: [String] -> String -> String -> IO Bool
-runChecks' modules funcSig body = if filterInf body then handleNotSupported $ do
-  arg <- (mapRandomArguments . parseTypeString) funcSig
-  result <- eval_ modules (fmtFunction_ body funcSig arg) 3000000
-  case result of
-    Nothing -> putStrLn "Timeout in running always-fail detection" >> return True
-    Just (Left err) -> putStrLn (displayException err) >> return False
-    Just (Right res) -> putStrLn res >> return True
-else pure $ not rejectInfiniteFunctions
-
-filterInf :: String -> Bool
-filterInf body = not $ any (`isInfixOf` body) excludeFunctions
-
+runChecks' modules funcSig body =
+  if hasNoInfiniteFuncs body
+  then handleNotSupported executeCheck
+  else pure $ not rejectInfiniteFunctions
+  where
+    executeCheck = do
+      arg <- generateTestInput modules (parseTypeString funcSig)
+      result <- eval_ modules (fmtFunction_ body funcSig arg) 3000000
+      case result of
+        Nothing -> putStrLn "Timeout in running always-fail detection" >> return True
+        Just (Left err) -> putStrLn (displayException err) >> return False
+        Just (Right res) -> putStrLn res >> return True
+
+hasNoInfiniteFuncs :: String -> Bool
+hasNoInfiniteFuncs body = not $ any (`isInfixOf` body) excludeFunctions
+
+-- *WIP: not used until the next iteration
 checkDuplicate :: [String] -> String -> String -> String -> IO Bool
 checkDuplicate modules funcSig bodyL bodyR = let fmt = fmtFunction_ in
-  if not (filterInf bodyL) || not (filterInf bodyR) then pure $ not rejectInfiniteFunctions 
+  if not (hasNoInfiniteFuncs bodyL) || not (hasNoInfiniteFuncs bodyR) then pure $ not rejectInfiniteFunctions
   else do
-    arg <- (mapRandomArguments . parseTypeString) funcSig
+    arg <- generateTestInput modules (parseTypeString funcSig)
     retL <- eval_ modules (fmt bodyL funcSig arg) 3000000
     retR <- eval_ modules (fmt bodyR funcSig arg) 3000000
 
     case (retL, retR) of
       (Just (Right l), Just (Right r)) -> putStrLn "duplicate fcn" >> return (l /= r)
       _ -> return False
 
+-- *WIP: not used until the next iteration
```

Comment:
I'd recommend making separate branches for your work. That is, one branch for this work, and another branch based on this one for next week's work.